### DISCUSSION TITLE
Using custom ARRAY subclass type is no longer possible

### DISCUSSION THREAD

**Discussion by kvesteri** [sqlalchemy/sqlalchemy#7476] [Usage Questions] [answered]
SQLAlchemy-Utils uses a custom PostgreSQL ARRAY subclass called CompositeArray. This class used to make it possible to have an array of composites in PostgreSQL. 

Consider the following declarative setup:

```python
import sqlalchemy as sa
from sqlalchemy_utils import CompositeAarray, CompositeType, CurrencyType


type_ = CompositeType(
    'money_type',
    [
        sa.Column('currency', CurrencyType),
        sa.Column('amount', sa.Integer)
    ]
)

class Account(Base):
    __tablename__ = 'account'
    id = sa.Column(sa.Integer, primary_key=True)
    balances = sa.Column(CompositeArray(type_))
```

For full reference test case see: https://github.com/kvesteri/sqlalchemy-utils/blob/master/tests/types/test_composite.py#L194-L209

SQLAlchemy version 1.4.26 changed the behaviour so that the custom CompositeArray is no longer used for bind parameter processing. The current CompositeArray internal implementation is very ugly and I'd like to change that (it overrides a private method called `_proc_array`). What is the recommended way in SA 1.4.26 for defining an array of composites?

**Comment by zzzeek** [comment, marked as answer]
Hi @kvesteri  -

I think the CompositeArray here might have stopped working for the psycopg2 driver, but it likely wasn't working already for the pg8000 and asyncpg drivers as we already were providing a dialect-level implementation for those two drivers.

anyway, this is a special case because looking at [CompositeArray](https://github.com/kvesteri/sqlalchemy-utils/blob/master/sqlalchemy_utils/types/pg_composite.py#L159) there's basically nothing there except a few lines that seem to try to alter the internal logic of the exisitng private _proc_array method https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/array.py#L298 .  I'm not able to detect visually exactly what changes with the change you are making since I need to see examples of what data comes back from different drivers, maybe there's already a list of tuples coming back?  I can't tell without watching it happen.   It looks like maybe this is effectively trying to get it to assume the "dim=1" behavior, rather than the "dim=None" behavior but I'm not sure.

if you can confirm that CompositeArray is not required as long as someone sends dimensions=1 to the type, you can just have your type set dimensions=1 at the constructor for now.  either way, please help us by illustrating a test case against pg.ARRAY that doesn't rely on sqlalchemy-utils, since this would become a unit test, where we shouldn't be calling list() up front. or where the detection needs to make different assumptions, and provide that as a bug report, that is, MCVE illustrating what it does wrong and as a bonus what we can adjust on our end, but if not we will work to figure that out, so that we can fix the bug on the SQLAlchemy side.      thanks!

**Reply by kvesteri** [reply]
Hi, your suggestion of having ARRAY(dimensions=1) solved the issue. Thanks!
